Enable L7 DNS proxy for nodes

atykhyy opened the pull request:
FQDN-based network policy is extremely useful for egress filtering, but is currently only supported in pods. This means that nodes must either rely on some other filtering mechanism, or resort to periodically updating L3/L4 CIDR policies. However, the code which collects DNS requests and combines them with policy to automatically produce L3/L4 filters is at endpoint level and is almost agnostic with respect to kind of endpoint (pod or node). This PR enables FQDN-based network policy for nodes by adding suitable iptables rules which redirect node's DNS requests to the same transparent proxy that is used for pods, and modifying the daemon startup sequence to avoid catch-22 between bootstrapping FQDN code and initialization of Kubernetes subsystem.

This PR works in simple cases (e.g. AKS cluster with default settings) but contains several kludges. I ask the community for advice and/or collaboration on these.

0. Node DNS traffic probably should be directed to the DNS proxy by the node's BPF program. I don't know BPF so I am adding rules to the OUTPUT chain of the mangle table (pkg/datapath/iptables/iptables.go:InstallHostDnsProxyRules) in what I hope is the correct place (pkg/endpoint/policy.go). These rules redirect node DNS packets to the proxy via the loopback interface using the same rule-based routing table used to redirect pod packets. To avoid infinite packet loops, I mark DNS proxy's forward requests as coming from proxy and also skip marking packets being redirected to the proxy as coming from host.
1. The host endpoint currently does not have any IP addresses associated with it. I could not figure out whether this was intentional, what was the correct place to source those addresses, and what would be the implications of associating IP addresses with the host endpoint, but this prevented existing DNS proxy code from allowing node's DNS requests, as it could not look up the source endpoint. Therefore I added a kludge which makes endpoint lookup work in the simple case of one node IPv4 address (daemon/cmd/fqdn.go:lookupEPByIP). This is also the reason why I added the `isHost bool` return value to the lookup callback and a special field for the restored host endpoint (fqdn/dnsproxy/proxy.go).
2. ~~Restored IP rules contain a string representation of the CIDR identifying the DNS server and don't match the destination IP address (fqdn/dnsproxy/proxy.go). It may be better to save/restore a richer representation of IP rules.~~ (Fixed.)
3. ~~Enabling L7 DNS rules for nodes prevents upgrades of the agent image from working, as the OUTPUT mangle rules redirecting node DNS requests to the proxy are not flushed but the proxy is not operational. This can be worked around by pre-pulling the images while the agent is still working. This is not elegant but on the other hand allowing any DNS traffic while the agent is not up would be a security hole. Agent pod restarts work thanks to restored IP rules.~~ (Struck out because [Cilium upgrade guide](https://docs.cilium.io/en/stable/operations/upgrade/) already requires a pre-flight check that pre-pulls agent images.)

FQDN-based network policy for nodes is likely to be a good addition to Cilium, as it would enable egress filtering by FQDNs for the whole cluster without relying on third-party solutions. Also, the same approach may possibly work for other L7 rules.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -623,6 +623,10 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+		if err := k8s.Init(option.Config); err != nil {
+			return nil, nil, fmt.Errorf("Unable to initialize Kubernetes subsystem: %w", err)
+		}
+
```
Why does this need to be moved? Maybe when you get to the point of submitting this for full review, you could separate this change out into its own commit and include a description about the motivation for moving this piece in the commit message.

atykhyy replied:
Sure, I'll split it out. The reason for moving this is because initializing Kubernetes needs to access kube-apiserver and it's often given as a FQDN, so DNS proxy has to be up before Kubernetes is initialized.

joestringer replied:
I see. One of the concerns I would see is that above, there is `d.k8sWatcher = watchers.NewK8sWatcher(...)` which seems like it may depend on `k8s.Init()`, so I don't know whether this change by itself is safe. I presume that the CI will find any issues though when we run it. Feel free to reach out here or on Slack when you think the PR is ready for integration testing.

atykhyy replied:
FWIW I went over the daemon startup sequence when I was moving this, and from what I could see `d.k8sWatcher = ...` and the following lines just create data structures. The watchers themselves are started later within an `if k8s.IsEnabled()` block.

joestringer replied:
:+1: ok great. We've been burnt a couple of times in the past, so I just wanted to make sure you're aware of the potential impact. Sounds like you're already ahead of me :)

atykhyy replied:
Split out as 3ea3c7664b8898f5547f34b970be2b2f6d362a34.

joestringer reviewed (commented):
Just noticed one thing while glancing through, but I won't do a full review just yet since we're discussing through another maybe-needed bugfix and this is still marked as draft :)

atykhyy committed:
Move k8s initialization after DNS proxy startup

kube-apiserver is often specified as an FQDN, thus DNS must be operational
in order for k8s subsystem to initialize successfully. If L7 DNS rules are
in effect on a node, this means that DNS proxy has to be up with restored
DNS rules. This commit moves k8s initialization in the agent to the earliest
point after DNS proxy has been started.

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>

atykhyy committed:
Basic support for looking up host endpoint by IP

The host endpoint currently does not have any IP addresses associated with it.
This prevents DNS proxy from working when source is the host endpoint.
This commit is a minimal change which enables the host endpoint to be looked up
by its IP address without touching any other code. The fix assumes that the host
has only one IP address.

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>

atykhyy committed:
Support CIDRs in restored DNS rules properly

Restored DNS rules code expected that LookupIPsBySecID returns only IP addresses,
but if an L7 DNS rule has toCIDR selectors configured, it also returns CIDRs,
and after such a DNS rule is restored, it does not match any target DNS servers
and the DNS proxy rejects DNS queries until the endpoint is regenerated.

This commit fixes this issue, replacing string IP addresses in restore.DNSRules
with a wrapped netip.Prefix (preserving JSON serialization of restored rules).

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>

atykhyy committed:
Enable L7 DNS proxy for nodes

FQDN-based network policy is extremely useful for egress filtering, but is
currently only supported in pods. This means that nodes must either rely on
some other filtering mechanism, or resort to periodically updating L3/L4 CIDR
policies. However, the code which collects DNS requests and combines them with
policy to automatically produce L3/L4 filters is at endpoint level and is
almost agnostic with respect to kind of endpoint (pod or node). This changeset
enables FQDN-based network policy for nodes by adding suitable iptables rules
which redirect node's DNS requests to the same transparent proxy that is used
for pods.

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>

jrajahalme commented on pkg/fqdn/restore/restore.go:
```diff
@@ -22,7 +25,70 @@ type IPRules []IPRule
+	if i := bytes.IndexByte(b, byte('/')); i < 0 {
+		var addr netip.Addr
+		if err = addr.UnmarshalText(b); err == nil {
+			*ip = RuleIPOrCIDR(netip.PrefixFrom(addr, 0xff))
```
Did you consider using the actual number of bits here (e.g., 32 for IPv4, 128 for IPv6)? That would allow use of `IsSingleIP()` instead of relying on `Bits()` returning `-1` in `IsAddr()` (could rename that as `IsSingleIP()` as well)?

Or is the intention to keep separation between a bare address and a single address CIDR? (e.g., 192.168.1.1 and 192.168.1.1/32)?

atykhyy replied:
> Did you consider using the actual number of bits here (e.g., 32 for IPv4, 128 for IPv6)? Or is the intention to keep separation between a bare address and a single address CIDR? (e.g., 192.168.1.1 and 192.168.1.1/32)?

I did consider that but rejected it, because if I understood the code correctly ipcache also distinguishes between a bare address and a CIDR, [e.g.](https://github.com/cilium/cilium/blob/master/pkg/ipcache/ipcache.go#L301):

	// Endpoint IP identities take precedence over CIDR identities, so if the
	// IP is a full CIDR prefix and there's an existing equivalent endpoint IP,
	// don't notify the listeners.

jrajahalme commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -178,8 +178,30 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
+	dest, err := restore.ParseRuleIPOrCIDR(destIP)
+	if err != nil || !dest.IsAddr() {
+		return false
+	}
```
This returns false if `dest` is a CIDR rather than an address. This seems unintentional, as below the lookup is done with both `dest` and `dest.ToSingleCIDR()`?

atykhyy replied:
It's an over-cautious sanity check. `destIP` being a packet's destination it is always an IP address, never a CIDR.

```
destAddr, err := netip.ParseAddr(destIP)
if err != nil {
    return false
}
dest := restore.RuleIPOrCIDR(netip.PrefixFrom(destAddr, 0xff)) // should introduce a function for this
```

should express the intent better. As an aside, with the new immutable types in `net/netip` the use of strings could be phased out all over the place, but it's a huge amount of work.

atykhyy replied:
The lookup is done with both `dest` and `dest.ToSingleCIDR()` because the set of rule IPs can contain both bare addresses (such as from resolution of `toEndpoint` rules) and CIDRs. I do these two checks as a fast lookup that covers most use cases before looping all over the set and (usually) rejecting the request.

jrajahalme reviewed (changes requested):
Thanks for this contribution :-) I've added some questions to some details below. The change to install an iptables rule to trigger proxy redirection does not seem the right approach for Cilium datapath. Datapath performs redirection based on the presence of a non-zero proxy port in the endpoint's policy map. So, instead of adding a (host) endpoint and destination port based iptables rule we'd need to fix whatever is preventing proxy redirection from `bpf/bpf_host.c` and `bpf/lib/host_firewall.h` in the first place. It looks like you have solved part of this difficulty by marking the outgoing traffic from the DNS proxy, as that mark is needed to not redirect traffic from the proxies again. 

A related consideration is that traffic from the L7 proxies that originated from a local (egress) or a remote (ingress) pod should not be subjected to the host firewall policy rules (L3/4 nor L7). I don't know if traffic from/to the local L7 proxies (indicated with some socket mark bits in `MagicMarkProxyMask` set) are excluded from the Host firewall enforcement or not. IMO they should be excluded.
